### PR TITLE
JAMES-2366 Group prefix

### PR DESCRIPTION
**Pull request by chibenwa**
Re-open of https://github.com/linagora/james-project/pull/1368 with only specific commits.

### CODE REVIEW

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java [hunk 1]
```diff
@@ -40,7 +40,12 @@
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Lists;
 
-public class MappingsImpl implements Mappings, Serializable {
+public class
+
+
+
+
+MappingsImpl implements Mappings, Serializable {
```
Comment (rouazana) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java]: gnii?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 2]
```diff
@@ -43,17 +43,21 @@ static Type detectType(String input) {
...
+        if (input.startsWith(Type.Group.asPrefix())) {
+            return Type.Group;
+        }
         return Type.Address;
     }
 
     Optional<MailAddress> asMailAddress();
 
     enum Type {
-        Regex("regex:", 3),
```
Comment (mbaechler) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: These orders don't make sense anymore.
My take is that we try to encode a policy into a sorting algorithm and having an int is convenient for sorting things.
Unfortunately the intent is very hard to guess.
Could we just have a Policy class ?

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 3]
```diff
@@ -167,6 +167,16 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
         return MappingsImpl.empty();
     }
 
+    private Mapping toMapping(String mappedAddress, Type type) {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: it doesn't make sense to me

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 4]
```diff
@@ -167,6 +167,16 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
...
+        switch (type) {
+            case Forward:
+            case Group:
+                return MappingImpl.of(type, mappedAddress);
+            default:
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: we should _never_ use default but list all enum values everytime to ensure a new type is not handled by accident.

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 5]
```diff
@@ -179,7 +189,7 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
             case Domain:
                 return Optional.of(user + "@" + Type.Domain.withoutPrefix(target));
             default:
-                return Optional.ofNullable(target);
+                return Optional.of(type.withoutPrefix(target));
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: same thing about default

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java [hunk 6]
```diff
@@ -108,7 +111,7 @@ public String getErrorMessage() {
 
     @Override
     public Optional<MailAddress> asMailAddress() {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingImpl.java]: please use a switch

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java (hunk 1 above)
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java]: oops

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 7]
```diff
@@ -308,6 +308,27 @@ public void removeForwardMapping(String user, Domain domain, String address) thr
         removeMapping(user, domain, mapping);
     }
 
+    @Override
+    public void addGroupMapping(String user, Domain domain, String address) throws RecipientRewriteTableException {
+        Mapping mapping = MappingImpl.group(address)
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: copy/paste for this seems wrong as it's likely that someone change this code without changing it in remove function, right ?

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 8]
```diff
@@ -144,6 +144,10 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
                         // Check if the returned mapping is the same as the
                         // input. If so return null to avoid loops
                         if (userName.equalsIgnoreCase(user) && targetDomain.equals(domain)) {
+                            if (type.equals(Type.Forward)) {
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: oh yes, let's add a if and a continue in this very simple method :trollface:

Comment (chibenwa) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: I did not find other solutions, sorry :trollface: 

Get a try, I think I spent 2 or 3 days on that. It is not easy as it seems.

**Comment by chibenwa** [comment]
```
I don't understand why we need to store group type at all.
Introducing group methods in the API would probably be enough to allow decoupling of the implementation, right ?
```

Well, what differentiates an address mapping (added from the cli for example) from a group mapping then?

REST API is not enough.

We discussed that for several weeks now, so I must admit I fdon't understand this remark, and I don't understand why it pops up so late.

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java (hunk 1 above)
Comment (chibenwa) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java]: Sorry.

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[90d313721299f353633e211ff59921d499cedeb2] Tests in error: 
[90d313721299f353633e211ff59921d499cedeb2]   ToRepositoryTest.userShouldBeAbleToAccessReprocessedMails:134 » ConditionTimeout
```

:green_apple:  locally so test this please

**Comment by chibenwa** [comment]
```
[90d313721299f353633e211ff59921d499cedeb2] Tests run: 16, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 75.246 sec <<< FAILURE! - in org.apache.james.webadmin.integration.WebAdminServerIntegrationTest
[90d313721299f353633e211ff59921d499cedeb2] addressGroupsEndpointShouldHandleRequests  Time elapsed: 3.116 sec  <<< FAILURE!
[90d313721299f353633e211ff59921d499cedeb2] java.lang.AssertionError: 
[90d313721299f353633e211ff59921d499cedeb2] 
[90d313721299f353633e211ff59921d499cedeb2] Expecting:
[90d313721299f353633e211ff59921d499cedeb2]   <[]>
[90d313721299f353633e211ff59921d499cedeb2] to contain only:
[90d313721299f353633e211ff59921d499cedeb2]   <["user1@domain.com", "user2@domain.com"]>
[90d313721299f353633e211ff59921d499cedeb2] but could not find the following elements:
[90d313721299f353633e211ff59921d499cedeb2]   <["user1@domain.com", "user2@domain.com"]>
[90d313721299f353633e211ff59921d499cedeb2] 
[90d313721299f353633e211ff59921d499cedeb2] 	at org.apache.james.webadmin.integration.WebAdminServerIntegrationTest.addressGroupsEndpointShouldHandleRequests(WebAdminServerIntegrationTest.java:295)
[90d313721299f353633e211ff59921d499cedeb2] 
[90d313721299f353633e211ff59921d499cedeb2] 
[90d313721299f353633e211ff59921d499cedeb2] Results :
[90d313721299f353633e211ff59921d499cedeb2] 
[90d313721299f353633e211ff59921d499cedeb2] Failed tests: 
[90d313721299f353633e211ff59921d499cedeb2]   WebAdminServerIntegrationTest.addressGroupsEndpointShouldHandleRequests:295 
[90d313721299f353633e211ff59921d499cedeb2] Expecting:
[90d313721299f353633e211ff59921d499cedeb2]   <[]>
[90d313721299f353633e211ff59921d499cedeb2] to contain only:
[90d313721299f353633e211ff59921d499cedeb2]   <["user1@domain.com", "user2@domain.com"]>
[90d313721299f353633e211ff59921d499cedeb2] but could not find the following elements:
[90d313721299f353633e211ff59921d499cedeb2]   <["user1@domain.com", "user2@domain.com"]>
```

Disapointing... I guess the added adresses don't have the `:group` prefix...

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java (hunk 2 above)
Comment (mbaechler) [reply server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: done

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java (hunk 1 above)
Comment (mbaechler) [reply server/data/data-library/src/main/java/org/apache/james/rrt/lib/MappingsImpl.java]: fixed

File: core/src/test/java/org/apache/james/core/UserTest.java [hunk 9]
```diff
@@ -78,7 +78,7 @@ public void fromLocalPartWithDomainStringVersionShouldThrowOnLocalPartThatContai
 
     @Test
     public void fromLocalPartWithDomainStringVersionShouldThrowOnNullDomainPart() {
-        assertThatThrownBy(() -> User.fromLocalPartWithDomain("local", null))
+        assertThatThrownBy(() -> User.fromLocalPartWithDomain("local", (String) null))
```
Comment (mbaechler) [inline core/src/test/java/org/apache/james/core/UserTest.java]: using a local variable `String domain = null` is a better solution

File: core/src/main/java/org/apache/james/core/User.java [hunk 10]
```diff
@@ -102,6 +104,11 @@ public String asString() {
             .orElse(localPart);
     }
 
+    public MailAddress asMailAddress() throws AddressException {
```
Comment (mbaechler) [inline core/src/main/java/org/apache/james/core/User.java]: throwing an exception here doesn't seem like a good idea. What about return `Optional<MailAddress>` instead ?

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 11]
```diff
@@ -128,39 +128,40 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
...
-        if (!maybeAddressWithMappingApplied.isPresent()) {
-            return Stream.empty();
-        }
-        String addressWithMappingApplied = maybeAddressWithMappingApplied.get();
-        LOGGER.debug("Valid virtual user mapping {}@{} to {}", user, domain.name(), addressWithMappingApplied);
+        ThrowingFunction<String, Stream<Mapping>> function =
+            (String stringMapping) -> convertAndRecurseMapping(associatedMapping.getType(), user, remainingLoops, stringMapping);
+
+        return associatedMapping.apply(user.asMailAddress())
+            .map(Throwing.function(function).sneakyThrow())
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: you don't need to wrap the ThrowingFunction in another ThrowingFunction IMO

File: server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java [hunk 12]
```diff
@@ -128,39 +128,40 @@ public Mappings getMappings(String user, Domain domain, int mappingLimit) throws
...
+            .orElse(Stream.empty());
+    }
 
+    private Stream<Mapping> convertAndRecurseMapping(Type mappingType, User originalUser, int remainingLoops, String addressWithMappingApplied) throws ErrorMappingException, RecipientRewriteTableException {
+        LOGGER.debug("Valid virtual user mapping {} to {}", originalUser, addressWithMappingApplied);
+
+        Stream<Mapping> possibleResult = Stream.of(toMapping(addressWithMappingApplied, mappingType));
```
Comment (mbaechler) [inline server/data/data-library/src/main/java/org/apache/james/rrt/lib/AbstractRecipientRewriteTable.java]: nonRecursiveMapping ?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 13]
```diff
@@ -52,19 +54,21 @@ static Type detectType(String input) {
...
+        Forward("forward:", 3, IdentityMappingBehaviour.ReturnIdentity),
+        Group("group:", 2, IdentityMappingBehaviour.Throw),
+        Address("", 4, IdentityMappingBehaviour.Throw);
 
         private final String asPrefix;
         private final int order;
+        private final IdentityMappingBehaviour identityMappingBehaviour;
 
-        Type(String asPrefix, Integer order) {
+        Type(String asPrefix, Integer order, IdentityMappingBehaviour identityMappingBehaviour) {
```
Comment (mbaechler) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: s/Behaviour/Policy ?

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 14]
```diff
@@ -83,6 +87,26 @@ public static boolean hasPrefix(String mapping) {
...
+        public IdentityMappingBehaviour getIdentityMappingBehaviour() {
+            return identityMappingBehaviour;
+        }
+    }
+
+    enum IdentityMappingBehaviour {
+        Throw,
+        ReturnIdentity;
+
+        public Stream<Mapping> handleIdentity(Stream<Mapping> mapping) {
```
Comment (mbaechler) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: it's way better to have an abstract method implemented by each enum value IMO.

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/UserRewritter.java [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+import org.apache.james.core.User;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.steveash.guavate.Guavate;
+import com.google.common.base.Splitter;
+import com.google.common.collect.ImmutableList;
+
+@FunctionalInterface
+public interface UserRewritter {
```
Comment (mbaechler) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/UserRewritter.java]: s/UserRewritter/UserRewriter

File: server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java [hunk 16]
```diff
@@ -57,21 +57,23 @@ static Type detectType(String input) {
     Optional<MailAddress> asMailAddress();
 
     enum Type {
-        Regex("regex:", 4, IdentityMappingBehaviour.Throw),
-        Domain("domain:", 1, IdentityMappingBehaviour.Throw),
-        Error("error:", 4, IdentityMappingBehaviour.Throw),
-        Forward("forward:", 3, IdentityMappingBehaviour.ReturnIdentity),
-        Group("group:", 2, IdentityMappingBehaviour.Throw),
-        Address("", 4, IdentityMappingBehaviour.Throw);
+        Regex("regex:", 4, IdentityMappingBehaviour.Throw, new UserRewritter.RegexRewriter()),
```
Comment (mbaechler) [inline server/data/data-api/src/main/java/org/apache/james/rrt/lib/Mapping.java]: there's no reuse of Rewriter (one type == on rewriter) so I would implement that as an abstract method of the enum instead.

**Comment by mbaechler** [comment]
replaced by https://github.com/linagora/james-project/pull/1378

File: core/src/main/java/org/apache/james/core/User.java [hunk 17]
```diff
@@ -47,7 +47,13 @@ public static User fromUsername(String username) {
...
 
-        return from(localPart,
+        return fromLocalPartWithDomain(localPart, Domain.of(domain));
+    }
+
+    public static User fromLocalPartWithDomain(String localPart, Domain domain) {
+        Preconditions.checkNotNull(domain);
+
+        return new User(localPart,
             Optional.of(domain));
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply core/src/main/java/org/apache/james/core/User.java]: ok

File: server/data/data-library/src/test/java/org/apache/james/rrt/lib/RecipientRewriteTableUtilTest.java [hunk 18]
```diff
@@ -28,10 +31,56 @@
...
+    public void regexMapShouldCorrectlyReplaceMatchingGroups() throws Exception {
+        MailAddress mailAddress = new MailAddress("prefix_abc@test");
+        assertThat(RecipientRewriteTableUtil.regexMap(mailAddress, MappingImpl.regex("prefix_(.*)@test:admin@${1}")))
+            .isEqualTo("admin@abc");
+    }
+
+    @Test
+    public void regexMapShouldCorrectlyReplaceSeveralMatchingGroups() throws Exception {
+        MailAddress mailAddress = new MailAddress("prefix_abc_def@test");
+        assertThat(RecipientRewriteTableUtil.regexMap(mailAddress, MappingImpl.regex("prefix_(.*)_(.*)@test:admin@${1}.${2}")))
```
[the remark this answers is not in the document]
Comment (mbaechler) [reply server/data/data-library/src/test/java/org/apache/james/rrt/lib/RecipientRewriteTableUtilTest.java]: ok

**Pull request closed without merging**